userspace/libsinsp/user.cpp

```diff
@@ -243,59 +282,110 @@ scap_userinfo *sinsp_usergroup_manager::add_user(const string &container_id, uin
 	scap_userinfo *usr = get_user(container_id, uid);
 	if (!usr)
 	{
-#ifdef HAVE_PWD_H
-		struct passwd *p = nullptr;
-		if (container_id.empty() && !name)
+		bool inserted{false};
+		if (name)
+		{
+			usr = userinfo_map_insert(
+				m_userlist[container_id],
+				uid,
+				gid,
+				name,
+				home,
+				shell);
+			inserted = true;
+		}
+		else if (container_id.empty())
 		{
+#ifdef HAVE_PWD_H
 			// On Host, try to load info from db
-			p = __getpwuid(uid);
+			auto* p = __getpwuid(uid);
 			if (p)
 			{
-				name = p->pw_name;
-				home = p->pw_dir;
-				shell = p->pw_shell;
+				usr = userinfo_map_insert(
+					m_userlist[container_id],
+					p->pw_uid,
+					p->pw_gid,
+					p->pw_name,
+					p->pw_dir,
+					p->pw_shell);
+				inserted = true;
 			}
-		}
 #endif
-
-		if (name == NULL)
-		{
-			name = "<NA>";
-		}
-		if (home == NULL)
-		{
-			home = "<NA>";
 		}
-		if (shell == NULL)
-		{
-			shell = "<NA>";
-		}
-
-		auto &userlist = m_userlist[container_id];
-		userlist[uid].uid = uid;
-		userlist[uid].gid = gid;
-		strlcpy(userlist[uid].name, name, MAX_CREDENTIALS_STR_LEN);
-		strlcpy(userlist[uid].homedir, home, SCAP_MAX_PATH_SIZE);
-		strlcpy(userlist[uid].shell, shell, SCAP_MAX_PATH_SIZE);
 
-		if (notify)
+		if (notify && inserted)
 		{
-			notify_user_changed(&userlist[uid], container_id);
+			notify_user_changed(usr, container_id);
 		}
-
-		usr = &userlist[uid];
 	}
-	else  if (name != NULL)
+	else if (name != NULL)
 	{
 		// Update user if it was already there
 		strlcpy(usr->name, name, MAX_CREDENTIALS_STR_LEN);
 		strlcpy(usr->homedir, home, SCAP_MAX_PATH_SIZE);
 		strlcpy(usr->shell, shell, SCAP_MAX_PATH_SIZE);
-
 	}
 	return usr;
 }
 
+scap_userinfo *sinsp_usergroup_manager::add_container_user(const std::string &container_id, int64_t pid, uint32_t uid, bool notify)
+{
+	ASSERT(!container_id.empty());
+	ASSERT(uid != 0);
+
+	auto userlist_it = m_userlist.find(container_id);
+	if(userlist_it != m_userlist.end())
+	{
+		// userlist for this container exists
+		auto it = userlist_it->second.find(uid);
+		// not an expected condition to miss, but handle it anyway
+		return (it == userlist_it->second.end())
+			       ? nullptr
+			       : &it->second;
+	}
+
+	scap_userinfo *retval{nullptr};
+
+#if defined HAVE_PWD_H && defined HAVE_FGET__ENT
+
+	if(false == m_ns_helper->in_own_ns_mnt(pid))
+	{
+		return nullptr;
```

Comment:
Might as well use it
```suggestion
		return retval;
```

Reply:
I choose to use `nullptr` for readability, to highlight we're not returning a value